Add Batcher Fallback: Channel Not Closed Test

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds a Test scenario for Batcher Fallback test consideration.  This PR adds a case where we have a multi-frame channel being submitted to the L1, and its submission is interrupted before the full channel can be sent.  In this scenario, we expect the Fallback Batcher to be able to pick up where the Espresso Batch left off.

<!-- ### This PR does not:  -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211892212379885

Ayiga committed:
Add test to check end of channel fallback

Asana task: https://app.asana.com/1/1208976916964769/project/1209976130071762/task/1211892212379885?focus=true

We need a test to check the fallback Batcher behavior in the event
that the Espresso Batcher is able to submit a partial Channel that is
im progress.

The specific scenario we want to test for is one concerning a
multi-frame channel that has had at least part of the full channel
submitted to the L1 by the Espresso Batcher, then no more.  After
which we swap to the Fallback Batcher, and we should be able to pick
up the missed / incomplete channel, and complete the transactions.

Ayiga committed:
Rename helper function to match naming pattern

Ayiga committed:
Fix lint issue with not checking error result of wait.For

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -55,3 +59,192 @@ func TestBatcherSwitching(t *testing.T) {
+// SendAsync is overridden to simulate a failure when shouldFail is true, and
+// to allow for one final transaction to be sent before failures begin when
+// triggerAfterOne is true.
+func (t *TxManagerIntercept) SendAsync(ctx context.Context, candidate txmgr.TxCandidate, ch chan txmgr.SendResponse) {
```
Why do we need Send and SendAsync?

Ayiga replied:
I was just implementing the interface for `TxManager` which declares both for different usage patterns.  In my testing it seems that only the `Send` method is utilized, but I just wanted to be certain and cover both use cases.

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -55,3 +59,192 @@ func TestBatcherSwitching(t *testing.T) {
+	// multiple frames.
+
+	// We want enough L2 Transactions to ensure we have multiple frames.
+	n := 10
```
Why is this enough?

Ayiga replied:
I don't have an answer for this.  This is just a value I attempted that triggered the intercept failure, so my assumption was that this was resulting in a multi-frame.

Upon further investigation this may not be the case, as I'm not sure how I can inspect the contents of an `L1` blocks to see if its contents consistent of a partial `L2` blocks or not.

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -55,3 +59,192 @@ func TestBatcherSwitching(t *testing.T) {
+	})
+	require.NoError(t, err)
+
+	if have, want := interceptTxManager.failureCount, 1; have < want {
```
Is this not redundant with the code starting at line  200?

Ayiga replied:
You are correct.

philippecamacho commented on espresso/environment/tx_helpers.go:
```diff
@@ -129,3 +130,39 @@ func RunSimpleL2Burn(ctx context.Context, t *testing.T, system *e2esys.System) {
+				opts.Nonce = nonce + uint64(i)
+				opts.Gas = 100_000
+
+				// opts.Data = make([]byte, 256)
```
Can we remove these commented lines?

philippecamacho replied:
Or maybe they should be uncommented?

Ayiga replied:
I initially attempted to add some arbitrary data to try and make the transactions larger.  However, this is a dynamic fee transaction, and this was resulting in an error, so I opted to just utilize simple transactions instead.

So you're correct in that unused commented out code should be removed.

philippecamacho reviewed (approved):
LGTM 

Left a few questions. Also added a fix to the Enclave CI test.

Ayiga committed:
Commit work in progress multi frame channel efforts

Ayiga committed:
Adjust settings to successfully trigger multi-frame channels

After a mob programming session @Quentinl was able to help identify a
a specific combiniation of parameters to successfully and consistently
trigger multi-frames within the Batcher.  This condition is a necessary
precusor to the test being attempted.

This commit updates the test with the information necessary to trigger
this condition and sets the necessary test criteria that we are aiming
to achieve.

Ayiga committed:
Perform some code cleanup

This change does a few things:

- Address linting issue causing CI failure
- Adjusts some golang forloop usage to be more modern
- Adjust function call signatures to remove unused variables

philippecamacho reviewed (changes requested):
Test is failing for me locally. Also Looks like there is a bug in `markFramesAsUnsuccessful`

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -55,3 +66,412 @@ func TestBatcherSwitching(t *testing.T) {
+	t.Lock()
+	defer t.Unlock()
+	for _, frame := range frames {
+		t.unsuccessfulFrames[frame.ID] = append(t.successfulFrames[frame.ID], frame)
```
Should not it be `t.unsuccessfulFrames[frame.ID] = append(t.unsuccessfulFrames[frame.ID], frame)` ?

Ayiga replied:
Good catch on this issue.  Though, I don't think would affect the test condition in any directly meaningful way, it should still be resolved.

Ayiga replied:
Resolved in 3cb092fcc3776dc059e28f52f0ec1fb1e3c012dd.

Ayiga committed:
Fix bug tracking unsuccessful frames in test

In the `TxManagerIntercept` there is a bug that appends the successful
frames to the unsuccessful ones.  While this bug isn't great in the
information that it taints, it doesn't actually have the large of an
impact on the test as a whole, as the resulting failure condition
would be triggered regardless.

This bug does affect the accurate tracking of failed frames which
could be valuable information for inspection.

Ayiga commented:
> Test is failing for me locally. Also Looks like there is a bug in `markFramesAsUnsuccessful`

Unless I am mistaken... the CI failure does **NOT** seem to be related to this test at all.  Instead it seems to be due to a timeout of a different test I believe.

philippecamacho commented on espresso/environment/e2e_helpers.go:
```diff
@@ -131,3 +133,184 @@ func WithL2BlockTime(blockTime time.Duration) E2eDevnetLauncherOption {
+}
+
+// WithBatcherCompressor is a configuration option that modifies the Compressor
+// setting of the Batcher Config.  This value will be utilzied to determine
```
```suggestion
// setting of the Batcher Config.  This value will be utilized to determine
```

philippecamacho commented:
> > Test is failing for me locally. Also Looks like there is a bug in `markFramesAsUnsuccessful`
> 
> Unless I am mistaken... the CI failure does **NOT** seem to be related to this test at all. Instead it seems to be due to a timeout of a different test I believe.

Yes, you are right. Here is the error message I have locally (I am on 3cb092fcc3776dc059e28f52f0ec1fb1e3c012dd). Here is the error I get:
```
INFO [01-05|13:54:56.271] Using L2 chain ID as namespace by default
    service.go:462:             INFO [01-05|13:54:56.271] Stopping batcher                         role=batcher
    service.go:508:             INFO [01-05|13:54:56.271] Batch Submitter stopped                  role=batcher
    14_batcher_fallback_test.go:368: 
        	Error Trace:	/home/leloup/repositories/espressosys/optimism-espresso-integration/espresso/environment/14_batcher_fallback_test.go:368
        	Error:      	Received unexpected error:
        	            	failed to setup batch submitter: failed to init Espresso: invalid Espresso config: attestation service URL is required when Espresso is enabled
        	Test:       	TestFallbackMechanismIntegrationTestChannelNotClosed
    setup.go:514: CLOSING
    service.go:308:             INFO [01-05|13:54:56.271] Stopping P
```

Ayiga commented:
> > > Test is failing for me locally. Also Looks like there is a bug in `markFramesAsUnsuccessful`
> > 
> > 
> > Unless I am mistaken... the CI failure does **NOT** seem to be related to this test at all. Instead it seems to be due to a timeout of a different test I believe.
> 
> Yes, you are right. Here is the error message I have locally (I am on [3cb092f](https://github.com/EspressoSystems/optimism-espresso-integration/commit/3cb092fcc3776dc059e28f52f0ec1fb1e3c012dd)). Here is the error I get:
> 
> ```
> INFO [01-05|13:54:56.271] Using L2 chain ID as namespace by default
>     service.go:462:             INFO [01-05|13:54:56.271] Stopping batcher                         role=batcher
>     service.go:508:             INFO [01-05|13:54:56.271] Batch Submitter stopped                  role=batcher
>     14_batcher_fallback_test.go:368: 
>         	Error Trace:	/home/leloup/repositories/espressosys/optimism-espresso-integration/espresso/environment/14_batcher_fallback_test.go:368
>         	Error:      	Received unexpected error:
>         	            	failed to setup batch submitter: failed to init Espresso: invalid Espresso config: attestation service URL is required when Espresso is enabled
>         	Test:       	TestFallbackMechanismIntegrationTestChannelNotClosed
>     setup.go:514: CLOSING
>     service.go:308:             INFO [01-05|13:54:56.271] Stopping P
> ```

This is the Environment variable stuff that Sneh added that I had brought up previously.  Are you running in `nix shell`?  Have you loaded the `.env` file that now seems to be required to run **ALL** e2e tests?

philippecamacho commented:
> > > > Test is failing for me locally. Also Looks like there is a bug in `markFramesAsUnsuccessful`
> > > 
> > > 
> > > Unless I am mistaken... the CI failure does **NOT** seem to be related to this test at all. Instead it seems to be due to a timeout of a different test I believe.
> > 
> > 
> > Yes, you are right. Here is the error message I have locally (I am on [3cb092f](https://github.com/EspressoSystems/optimism-espresso-integration/commit/3cb092fcc3776dc059e28f52f0ec1fb1e3c012dd)). Here is the error I get:
> > ```
> > INFO [01-05|13:54:56.271] Using L2 chain ID as namespace by default
> >     service.go:462:             INFO [01-05|13:54:56.271] Stopping batcher                         role=batcher
> >     service.go:508:             INFO [01-05|13:54:56.271] Batch Submitter stopped                  role=batcher
> >     14_batcher_fallback_test.go:368: 
> >         	Error Trace:	/home/leloup/repositories/espressosys/optimism-espresso-integration/espresso/environment/14_batcher_fallback_test.go:368
> >         	Error:      	Received unexpected error:
> >         	            	failed to setup batch submitter: failed to init Espresso: invalid Espresso config: attestation service URL is required when Espresso is enabled
> >         	Test:       	TestFallbackMechanismIntegrationTestChannelNotClosed
> >     setup.go:514: CLOSING
> >     service.go:308:             INFO [01-05|13:54:56.271] Stopping P
> > ```
> 
> This is the Environment variable stuff that Sneh added that I had brought up previously. Are you running in `nix shell`? Have you loaded the `.env` file that now seems to be required to run **ALL** e2e tests?

I am running nix yes but this `.env` file is new  to me. Our `README_ESPRESSO.md` does not seem to mention it. Could you update it please?

Ayiga commented:
> I am running nix yes but this `.env` file is new to me. Our `README_ESPRESSO.md` does not seem to mention it. Could you update it please?

I did not add this file, nor this requirement.  This was my point in bringing this up in the first place.  These newly added environment variables **SHOULD NOT** be required for testing or running tests.  It seems absurd to me that they have been added as a non-optional requirement without documentation.

Ayiga commented:
I think the best solution, in addition to documenting their requirements / values (which I may need help from @Sneh1999 in order to accomplish), would be to make these truely optional, especially in the `e2e` tests, by populating default values for them when they launch.

Additionally, this zk verifier likely does **NOT** need to be launched for every single `e2e` test.

philippecamacho commented:
CI looks stuck. Maybe we should resolve the conflicts and start it again.

Ayiga committed:
Update espresso/environment/e2e_helpers.go

Co-authored-by: Phil <philippe@espressosys.com>

Ayiga committed:
Replace Disable Batcher setting references

There are a number of places in our testing setup where we are
explicitly preventing the Batcher from starting on launch.  Instead of
rewriting this same option every time we want to use it, we should
reference a built in option that we can reference continually.  This
allows for non-repeated code and improved documentation as to the point
and purpose of this option.

Ayiga committed:
Refactor custom wait in test

There's a condition being waited on in the switch to fallback batcher
test.  This wait is useful, and can be reused between tests.  But the
wait itself is somewhat hiding it's intention by being inline defined
within the test itself.  We should pull this wait out so it can be
easily used, and its intention / purpose can be more easily documented.

Ayiga committed:
Cleanup code reuse in frame decoding

When decoding frame information for one of the Batcher fallback tests,
there are similar code paths taken that result in most of the code being
reused.  We should clean up this code reuse so that we don't repeat
ourselves in order to avoid diverging logic.  Additionally, it allows us
to reduce the amount of code needing to be maintained, and more clearly
document the intention of the code, and the consistency with how we
perform this frame decoding process.

Ayiga committed:
Relocate deferred stop calls

The Stop calls should occur as close to the launch of the environment as
possible.  As a result, any deferred calls to Stop for the system or the
Espresso Dev Node should occur as close to their occurence as possible.

Ayiga committed:
Modify Initial L2Verif wait to be longer

With the specific Frame and Channel settings being specified in the
`TestFallbackMechanismIntegrationTestChannelNotClosed` test, the initial
startup check for the L2 Verifier is failing.  This is due to our
settings requiring the Verifier process to take a bit longer than
normal.  In general, we want to give it more time, but the time frame
for the failure is hard-coded in the `wait` function being utilized.
While we **could** add a simple `time.Sleep`, and this would work, this
is generally a bad appraoch as it just adds an unchecked delay.
Instead, we opt to utilize a simple `retry` for up to `n` times.  In
this case, we only need to wait up to `3x` the normal time, so ensure
that we perform at least `3` times.

Ayiga committed:
Fix failure in Batcher Fallback test

The TestFallbackMechanismIntegrationTestChannelNotClosed test fails
locally without stopping, in spite of the overall time limit being
specified on the test.  After some troubleshooting and debugging, We
were able to chase down the cause to be due to the
`RunSimpleMultiTransactions`.  It's unclear as to why this was causing
the process to hang for as long as it was.  It seemed to not be handling
timeout errors well for some reason.

Either way, we fority this helper by setting an explicit time limit on
it, and referncing the context whenever we're performing channel
operations.  This should allow the channel operations themselves not to
block and hang the test.

After this modification we were able to determine that this process was
failing due to insufficient gas being provided.  For some reason when
running the transactions through this mechanism, we require even more
gas than we're normally need.  This seems a bit odd, perhaps it has to
do with the differences in the transaction construction.  In any case,
we up the gas being provided so that this becomes a non-issue.

Ayiga committed:
Fix linting issues

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -133,3 +152,425 @@ Loop:
+
+	if t.shouldFail {
+		t.failureCount++
+		t.markFramesAsSuccessful(frames)
```
Should not it be `t.markFramesAsUnSuccessful(frames)` like in `SendAsync`?

Ayiga replied:
You are correct!

Ayiga replied:
Fixed in a030ed3f96f0dfa42f778480d94da0a0a0eb260f.

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -133,3 +152,425 @@ Loop:
+		t.shouldFail = true
+	}
+
+	t.markFramesAsUnsuccessful(frames)
```
Similarly should not it be `t.markFramesAsSuccessful(frames)`.

Ayiga replied:
Also correct!

Ayiga replied:
Fixed in a030ed3f96f0dfa42f778480d94da0a0a0eb260f.

philippecamacho commented on espresso/environment/tx_helpers.go:
```diff
@@ -129,3 +130,51 @@ func RunSimpleL2Burn(ctx context.Context, t *testing.T, system *e2esys.System) {
+// RunSimpleMultiTransactions sends numTransactions simple L2 transactions
+// from Bob's account and returns the receipts.
+//
+// This is all attempted in porallel, as it will spawn a separate goroutine
```
```suggestion
// This is all attempted in parallel, as it will spawn a separate goroutine
```

Ayiga committed:
Update espresso/environment/tx_helpers.go

Co-authored-by: Phil <philippe@espressosys.com>

Ayiga committed:
Correct failure vs success in Send

The triggered conditions for failures and successes are backwards in the
`Send` method of `TxMangerIntercept`.  Their specific frame markers
should be switched.

philippecamacho commented on espresso/environment/14_batcher_fallback_test.go:
```diff
@@ -133,3 +152,425 @@ Loop:
+
+	// Let's make sure that the system is progressing initially for both
+	// the Sequencer, the Verifier, and the L1Node
+	// err = wait.ForNextBlock(ctx, l2Seq)
```
```suggestion
```

philippecamacho reviewed (approved):
LGTM

Ayiga committed:
Update espresso/environment/14_batcher_fallback_test.go

Co-authored-by: Phil <philippe@espressosys.com>
